doc/spacebar.1
'\" t
.\"     Title: spacebar
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.10
.\"      Date: 2021-04-08
.\"    Manual: spacebar manual
.\"    Source: spacebar
.\"  Language: English
.\"
.TH "SPACEBAR" "1" "2021-04-08" "spacebar" "spacebar manual"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
spacebar \- a simple status bar for macOS
.SH "SYNOPSIS"
.sp
\fBspacebar\fP [\fB\-v\fP,\fB\-\-version\fP|\fB\-V\fP,\fB\-\-verbose\fP|\fB\-m\fP,\fB\-\-message\fP \fImsg\fP|\fB\-c\fP,\fB\-\-config\fP \fIconfig_file\fP]
.SH "DESCRIPTION"
.sp
\fBspacebar\fP is a simple status bar for macOS.
.SH "OPTIONS"
.sp
\fB\-v\fP, \fB\-\-version\fP
.RS 4
Print the version and exit.
.RE
.sp
\fB\-V\fP, \fB\-\-verbose\fP
.RS 4
Output debug information to stdout.
.RE
.sp
\fB\-m\fP, \fB\-\-message\fP \fI<msg>\fP
.RS 4
Send message to a running instance of spacebar.
.RE
.sp
\fB\-c\fP, \fB\-\-config\fP \fI<config_file>\fP
.RS 4
Use the specified configuration file.
.RE
.SH "CONFIG"
.SS "General Syntax"
.sp
spacebar \-m config <setting>
.RS 4
Get or set the value of <setting>.
.RE
.SS "Settings"
.sp
\fBdebug_output\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable output of debug information to stdout.
.RE
.sp
\fBdisplay\fP [\fI<display>\fP]
.RS 4
Display to draw the bar on.
.br
Must be one of \fImain\fP or \fIall\fP.
.RE
.sp
\fBposition\fP [\fI<position>\fP]
.RS 4
Position on the screen to draw the bar.
.br
Must be one of \fItop\fP or \fIbottom\fP.
.RE
.sp
\fBheight\fP [\fI<NUM>\fP]
.RS 4
Height in pixels to draw the bar.
.RE
.sp
\fBtitle\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable focused window title display at the centre of the bar.
.br
This cannot be enabled when \fIcenter_shell\(cqis set to \(aqon\fP.
.RE
.sp
\fBspaces\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable space indicator on the left of the bar.
.RE
.sp
\fBspace_icon_strip\fP [\fI<sym_1> <sym_2> <sym_n>\fP]
.RS 4
Specify symbols separated by whitespace to be used for visualizing spaces.
.RE
.sp
\fBspaces_for_all_displays\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable space indicators for spaces on each display.
.RE
.sp
\fBspace_icon\fP [\fI<sym>\fP]
.RS 4
Specify a general symbol to use for any given space that does not have a match in \fIspace_icon_strip\fP.
.RE
.sp
\fBspace_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the current space icon.
.RE
.sp
\fBspace_icon_color_secondary\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the current space icon of the second display.
.RE
.sp
\fBspace_icon_color_tertiary\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the current space icon of the third display.
.RE
.sp
\fBdisplay_separator\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable separator between the spaces for each display.
.RE
.sp
\fBdisplay_separator_icon\fP [\fI<sym>\fP]
.RS 4
Specify a symbol to use for the display separator.
.RE
.sp
\fBdisplay_separator_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the display separator icon.
.RE
.sp
\fBclock\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable clock on the right of the bar.
.RE
.sp
\fBpower\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable power indicator on the right of the bar.
.RE
.sp
\fBpadding_left\fP [\fI<NUM>\fP]
.RS 4
Padding in pixels between the left edge of the display and the first item.
.RE
.sp
\fBpadding_right\fP [\fI<NUM>\fP]
.RS 4
Padding in pixels between the right edge of the display and the last item.
.RE
.sp
\fBspacing_left\fP [\fI<NUM>\fP]
.RS 4
Spacing in pixels between the left space indicators.
.RE
.sp
\fBspacing_right\fP [\fI<NUM>\fP]
.RS 4
Spacing in pixels between the right status segments.
.RE
.sp
\fBtext_font\fP [\fI<font_family>:<font_style>:<font_size>\fP]
.RS 4
Specify name, style and size of font to use for drawing text.
.br
Use \fIFont Book.app\fP to identify the correct name.
.RE
.sp
\fBicon_font\fP [\fI<font_family>:<font_style>:<font_size>\fP]
.RS 4
Specify name, style and size of font to use for drawing icon symbols.
.br
Use \fIFont Book.app\fP to identify the correct name.
.RE
.sp
\fBbackground_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing status bar background.
.RE
.sp
\fBforeground_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing status bar elements.
.RE
.sp
\fBpower_icon_strip\fP [\fI<sym_battery> <sym_ac>\fP]
.RS 4
Specify two symbols separated by whitespace.
.br
The first symbol represents battery power and the second symbol indicates AC.
.RE
.sp
\fBpower_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the power (charging) icon.
.RE
.sp
\fBbattery_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the battery icon.
.RE
.sp
\fBclock_icon\fP [\fI<sym>\fP]
.RS 4
Specify a symbol to represent the current time.
.RE
.sp
\fBclock_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the clock icon.
.RE
.sp
\fBclock_format\fP [\fI<sym>\fP]
.RS 4
Specify a format for the current time, according to the strftime function.
.RE
.sp
\fBdnd\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable a DoNotDisturb indicator on the right of the bar.
.RE
.sp
\fBdnd_icon\fP [\fI<sym>\fP]
.RS 4
Specify a symbol to represent the current DoNotDisturb status.
.RE
.sp
\fBdnd_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the DoNotDisturb icon.
.RE
.sp
\fBleft_shell\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable shell output on the left of the bar.
.RE
.sp
\fBright_shell\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable shell output on the right of the bar.
.RE
.sp
\fBcenter_shell\fP [\fI<BOOL_SEL>\fP]
.RS 4
Enable shell output at the center of the bar.
.br
This cannot be enabled when \fItitle\fP is set to \fIon\fP.
.RE
.sp
\fBleft_shell_icon\fP [\fI<sym>\fP]
.RS 4
Specify a symbol to prefix the left shell output.
.RE
.sp
\fBleft_shell_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the left shell icon.
.RE
.sp
\fBleft_shell_command\fP [\fI<shell command>\fP]
.RS 4
Command pipeline to retrieve the output for displaying in the left shell section.
.br
There is NO timeout protection for the command pipeline, so be sure to set it to something that returns output quickly.
.RE
.sp
\fBright_shell_icon\fP [\fI<sym>\fP]
.RS 4
Specify a symbol to prefix the right shell output.
.RE
.sp
\fBright_shell_icon_color\fP [\fI<COLOR>\fP]
.RS 4
Color to use for drawing the right shell icon.
.RE
.sp
\fBright_shell_command\fP [\fI<shell command>\fP]
.RS 4
Command pipeline to retrieve the output for displaying in the right shell section.
.br
There is NO timeout protection for the command pipeline, so be sure to set it to something that returns output quickly.
.RE
.sp
\fBcenter_shell_command\fP [\fI<shell command>\fP]
.RS 4
Command pipeline to retrieve the output for displaying in the center shell section.
.br
There is NO timeout protection for the command pipeline, so be sure to set it to something that returns output quickly.
.br
.RE
.SH "EXIT CODES"
.sp
If \fBspacebar\fP can\(cqt handle a message, it will return a non\-zero exit code.
.SH "AUTHOR"
.sp
Calum MacRae <hi at cmacr.ae>

CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased](https://github.com/cmacrae/spacebar/compare/v1.2.1...HEAD)

**Changed**
- Fixed a bug introduced in Big Sur where buffer reads were incorrect
- Improved efficiency of bar's initialisation
- Fixed DoNotDisturb indicator on Big Sur

**Added**
- New `left|center|right` shell sections: display custom text based on shell pipelines
- Option to turn focused window title display on or off (thanks [@Norviah](https://github.com/Norviah)!)
- Option to turn the spaces indicator on or off
- Option to turn the clock on or off
- Option to turn the power indicator on or off
- Option to turn the DoNotDisturb indicator on or off
- Option to set the padding between the first/last item and the left/right edge of the display
- Default to "Solid" style of Font Awesome 5 Free icon font
- Option to display spaces for all displays, including: optional separator, secondary & tertiary space indicator colours in relation to display
- Option to specify only drawing one bar on the main display or multiple bars, one for each display (this aligns with yabai's `external_bar` option)
- Provide a [Nix Flake](https://nixos.wiki/wiki/Flakes)

## [1.2.1](https://github.com/cmacrae/spacebar/releases/tag/v1.2.1) - 2020-11-18

**Changed**
- Fixed a bug where querying for the value of `space_icon_strip` would set it to default (thanks [@jraregris](https://github.com/jraregris))

## [1.2.0](https://github.com/cmacrae/spacebar/releases/tag/v1.2.0) - 2020-11-18

**Added**
- Spacing configuration options

## [1.1.1](https://github.com/cmacrae/spacebar/releases/tag/v1.1.1) - 2020-07-21

**Changed**
- Fixed a bug where long window titles would draw over the right status area
- Padding between items in the status area based on current values

## [1.1.0](https://github.com/cmacrae/spacebar/releases/tag/v1.1.0) - 2020-07-17

**Added**
- Height configuration option

## [1.0.0](https://github.com/cmacrae/spacebar/releases/tag/v1.0.0) - 2020-07-16

**Added**
- Option to position at the top or bottom of the screen
- Individual colour settings for each icon in the right strip (`dnd`, `power`, `clock`)
- DoNotDisturb indicator

**Changed**
- Current space indicated by colouring the glyph
- Removal of underlines
- Fixed flicker bug when changing monitor focus (thanks [@tom-auger](https://github.com/tom-auger))

## Pre-1.0.0
This changelog was not kept up to date prior to `1.0.0`.  
See the commit log for more information.


src/spacebar.c
#define SOCKET_PATH_FMT         "/tmp/spacebar_%s.socket"
#define LCFILE_PATH_FMT         "/tmp/spacebar_%s.lock"

#define CLIENT_OPT_LONG         "--message"
#define CLIENT_OPT_SHRT         "-m"

#define DEBUG_VERBOSE_OPT_LONG  "--verbose"
#define DEBUG_VERBOSE_OPT_SHRT  "-V"
#define VERSION_OPT_LONG        "--version"
#define VERSION_OPT_SHRT        "-v"
#define CONFIG_OPT_LONG         "--config"
#define CONFIG_OPT_SHRT         "-c"

#define MAJOR 1
#define MINOR 2
#define PATCH 1

extern int SLSMainConnectionID(void);

#define CONNECTION_CALLBACK(name) void name(uint32_t type, void *data, size_t data_length, void *context, int cid)
typedef CONNECTION_CALLBACK(connection_callback);
extern CGError SLSRegisterConnectionNotifyProc(int cid, connection_callback *handler, uint32_t event, void *context);

struct event_loop g_event_loop;
void *g_workspace_context;
struct process_manager g_process_manager;
struct display_manager g_display_manager;
struct application_manager g_application_manager;
struct daemon g_daemon;
struct bar_manager g_bar_manager;
int g_connection;

char g_socket_file[MAXLEN];
char g_config_file[4096];
char g_lock_file[MAXLEN];
bool g_verbose;

static int client_send_message(int argc, char **argv)
{
    if (argc <= 1) {
        error("spacebar-msg: no arguments given! abort..\n");
    }

    char *user = getenv("USER");
    if (!user) {
        error("spacebar-msg: 'env USER' not set! abort..\n");
    }

    int sockfd;
    char socket_file[MAXLEN];
    snprintf(socket_file, sizeof(socket_file), SOCKET_PATH_FMT, user);

    if (!socket_connect_un(&sockfd, socket_file)) {
        error("spacebar-msg: failed to connect to socket..\n");
    }

    int message_length = argc;
    int argl[argc];

    for (int i = 1; i < argc; ++i) {
        argl[i] = strlen(argv[i]);
        message_length += argl[i];
    }

    char message[message_length];
    char *temp = message;

    for (int i = 1; i < argc; ++i) {
        memcpy(temp, argv[i], argl[i]);
        temp += argl[i];
        *temp++ = '\0';
    }
    *temp++ = '\0';

    if (!socket_write_bytes(sockfd, message, message_length)) {
        error("spacebar-msg: failed to send data..\n");
    }

    shutdown(sockfd, SHUT_WR);

    int result = EXIT_SUCCESS;
    int byte_count = 0;
    char rsp[BUFSIZ];

    struct pollfd fds[] = {
        { sockfd, POLLIN, 0 }
    };

    while (poll(fds, 1, -1) > 0) {
        if (fds[0].revents & POLLIN) {
            if ((byte_count = recv(sockfd, rsp, sizeof(rsp)-1, 0)) <= 0) {
                break;
            }

            rsp[byte_count] = '\0';

            if (rsp[0] == FAILURE_MESSAGE[0]) {
                result = EXIT_FAILURE;
                fprintf(stderr, "%s", rsp + 1);
                fflush(stderr);
            } else {
                fprintf(stdout, "%s", rsp);
                fflush(stdout);
            }
        }
    }

    socket_close(sockfd);
    return result;
}

static void acquire_lockfile(void)
{
    int handle = open(g_lock_file, O_CREAT | O_WRONLY, 0600);
    if (handle == -1) {
        error("spacebar: could not create lock-file! abort..\n");
    }

    struct flock lockfd = {
        .l_start  = 0,
        .l_len    = 0,
        .l_pid    = getpid(),
        .l_type   = F_WRLCK,
        .l_whence = SEEK_SET
    };

    if (fcntl(handle, F_SETLK, &lockfd) == -1) {
        error("spacebar: could not acquire lock-file! abort..\n");
    }
}

static bool get_config_file(char *restrict filename, char *restrict buffer, int buffer_size)
{
    char *xdg_home = getenv("XDG_CONFIG_HOME");
    if (xdg_home && *xdg_home) {
        snprintf(buffer, buffer_size, "%s/spacebar/%s", xdg_home, filename);
        if (file_exists(buffer)) return true;
    }

    char *home = getenv("HOME");
    if (!home) return false;

    snprintf(buffer, buffer_size, "%s/.config/spacebar/%s", home, filename);
    if (file_exists(buffer)) return true;

    snprintf(buffer, buffer_size, "%s/.%s", home, filename);
    return file_exists(buffer);
}

static void exec_config_file(void)
{
    if (!*g_config_file && !get_config_file("spacebarrc", g_config_file, sizeof(g_config_file))) {
        notify("configuration", "could not locate config file..");
        return;
    }

    if (!file_exists(g_config_file)) {
        notify("configuration", "file '%s' does not exist..", g_config_file);
        return;
    }

    if (!ensure_executable_permission(g_config_file)) {
        notify("configuration", "could not set the executable permission bit for '%s'", g_config_file);
        return;
    }

    if (!fork_exec(g_config_file, NULL)) {
        notify("configuration", "failed to execute file '%s'", g_config_file);
        return;
    }
}

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated-declarations"
static inline void init_misc_settings(void)
{
    char *user = getenv("USER");
    if (!user) {
        error("spacebar: 'env USER' not set! abort..\n");
    }

    snprintf(g_socket_file, sizeof(g_socket_file), SOCKET_PATH_FMT, user);
    snprintf(g_lock_file, sizeof(g_lock_file), LCFILE_PATH_FMT, user);

    NSApplicationLoad();
    signal(SIGCHLD, SIG_IGN);
    signal(SIGPIPE, SIG_IGN);
    CGSetLocalEventsSuppressionInterval(0.0f);
    CGEnableEventStateCombining(false);
    g_connection = SLSMainConnectionID();
}
#pragma clang diagnostic pop

static CONNECTION_CALLBACK(connection_handler)
{
}

static void parse_arguments(int argc, char **argv)
{
    if ((string_equals(argv[1], VERSION_OPT_LONG)) ||
        (string_equals(argv[1], VERSION_OPT_SHRT))) {
        fprintf(stdout, "spacebar-v%d.%d.%d\n", MAJOR, MINOR, PATCH);
        exit(EXIT_SUCCESS);
    }

    if ((string_equals(argv[1], CLIENT_OPT_LONG)) ||
        (string_equals(argv[1], CLIENT_OPT_SHRT))) {
        exit(client_send_message(argc-1, argv+1));
    }

    for (int i = 1; i < argc; ++i) {
        char *opt = argv[i];

        if ((string_equals(opt, DEBUG_VERBOSE_OPT_LONG)) ||
            (string_equals(opt, DEBUG_VERBOSE_OPT_SHRT))) {
            g_verbose = true;
        } else if ((string_equals(opt, CONFIG_OPT_LONG)) ||
                   (string_equals(opt, CONFIG_OPT_SHRT))) {
            char *val = i < argc - 1 ? argv[++i] : NULL;
            if (!val) error("spacebar: option '%s|%s' requires an argument!\n", CONFIG_OPT_LONG, CONFIG_OPT_SHRT);
            snprintf(g_config_file, sizeof(g_config_file), "%s", val);
        } else {
            error("spacebar: '%s' is not a valid option!\n", opt);
        }
    }
}

int main(int argc, char **argv)
{
    if (argc > 1) {
        parse_arguments(argc, argv);
    }

    if (is_root()) {
        error("spacebar: running as root is not allowed! abort..\n");
    }

    if (!ax_privilege()) {
        error("spacebar: could not access accessibility features! abort..\n");
    }

    init_misc_settings();
    acquire_lockfile();

    if (!event_loop_init(&g_event_loop)) {
        error("spacebar: could not initialize event_loop! abort..\n");
    }

    process_manager_init(&g_process_manager);
    workspace_event_handler_init(&g_workspace_context);
    application_manager_init(&g_application_manager);
    bar_manager_init(&g_bar_manager);

    event_loop_begin(&g_event_loop);
    display_manager_begin(&g_display_manager);
    process_manager_begin(&g_process_manager);
    workspace_event_handler_begin(&g_workspace_context);
    application_manager_begin(&g_application_manager);
    bar_manager_begin(&g_bar_manager);
    SLSRegisterConnectionNotifyProc(g_connection, connection_handler, 1204, NULL);

    if (!socket_daemon_begin_un(&g_daemon, g_socket_file, message_handler)) {
        error("spacebar: could not initialize daemon! abort..\n");
    }

    exec_config_file();
    CFRunLoopRun();
    return 0;
}
